Document single op replacement

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50116 Document single op replacement**

Differential Revision: [D25803457](https://our.internmc.facebook.com/intern/diff/D25803457)

Commit:
Document single op replacement

[ghstack-poisoned]

Commit:
Update on "Document single op replacement"




[ghstack-poisoned]

Commit:
Update on "Document single op replacement"




[ghstack-poisoned]

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,58 @@
+# on the `target` attribute. In this particular case, we have two
+# different representations of addition, so we'll use a regex to match
+# on `target`. (If you want to replace a `torch`-specific operator,
+# you can match on a simple string.)
```
I think the things we should be matching on are a bit different, e.g.:

```
import torch
import torch.fx

class Additions(torch.nn.Module):
    def forward(self, x, y):
        return x + y, torch.add(x, y), x.add(y)

traced = torch.fx.symbolic_trace(Additions())
print(traced.graph)
"""
graph(x, y):
    %add_1 : [#users=1] = call_function[target=<built-in function add>](args = (%x, %y), kwargs = {})
    %add_2 : [#users=1] = call_function[target=<built-in method add of type object at 0x7f860323f7a0>](args = (%x, %y), kwargs = {})
    %add_3 : [#users=1] = call_method[target=add](args = (%x, %y), kwargs = {})
    return (add_1, add_2, add_3)
"""
```

the cases are:
1. call_function `<built-in function add>`. This is actually `operator.add` so we can match on equality with that function directly
2. call_function `<built-in method add of type object at 0x7f860323f7a0>` like you mentioned this is `torch.add`, which we can similarly match directly
3. This is the Tensor method call, so we can match target as a string

I think it's best do to:

```
patterns = [
    ('call_function', operator.add),
    ('call_function', torch.add),
    ('call_method', 'add')
]
```

Then later we can match like:

```
n : Node = ...
to_match = (n.op, n.target)
if any(to_match == pattern for pattern in patterns):
  # is a match, can swap stuff out
```

This should be safer than doing a string regex match.

BTW the operator form actually has an additional level of dispatch in the Python interpreter that goes through the string cat/addition and `__add__`, `__radd__` dispatch chain. It might be worth mentioning that the user should make sure that both operands are supposed to be tensors. (We don't have a way to confirm that directly, it's part of my operator normalization proposal doc)

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,58 @@
+        # Create a replacement node with the new op
+        new_node = traced.graph.call_function(torch.bitwise_and, n.args, n.kwargs)
+        # Move the new node to the correct spot
+        n.append(new_node)
```
Can we just use `graph.inserting_after`?

jamesr66a reviewed (commented):
Nice!

Commit:
Update on "Document single op replacement"


Differential Revision: [D25803457](https://our.internmc.facebook.com/intern/diff/D25803457)

[ghstack-poisoned]

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,64 @@
+on the node's ``target`` attribute).
+3. Create a replacement node.
+4. Use the fx built-in ``replace_all_uses_with`` to replace the current
+node with the replacement.
```
Probably want to mention that this replaces the nodes "uses" rather than replacing the node itself. The current wording is ambiguous

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,64 @@
+Python code to reflect the new Graph state.
+
+Currently, FX does not provide operator normalization. It's up to the
+user to confirm a way to confirm that any replacement operators will
```
`to confirm` seems to show up twice in this sentence. Also, a user reading this is probably not going to know what "operator normalization" means, so probably better to add a clause or example explaining the implication

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,64 @@
+# to denote addition. The possible cases are:
+#     1. `x + y` - A `call_function` Node with target
+#        `<built-in function add>`. This is `operator.add`, so we can
+#         match on equality with that function directly.
```
Remind me to put out a patch that makes the printout of builtin functions and operators more friendly if I forget. The python repr `<built-in function... >` is super confusing. I'll try to get something out today

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,64 @@
+for n in traced.graph.nodes:
+    # If the target matches one of the patterns
+    to_match = (n.op, n.target)
+    if any(to_match == pattern for pattern in patterns):
```
Not necessary to do, but I realized after I wrote that original comment that `patterns` can also be a set since all the items are hashable

jamesr66a commented on torch/fx/examples/replace_op.py:
```diff
@@ -0,0 +1,64 @@
+            new_node = traced.graph.call_function(torch.bitwise_and, n.args, n.kwargs)
+            n.replace_all_uses_with(new_node)
+        # Remove the old node from the graph
+        traced.graph.erase_node(n)
```
Another cool thing to add might be some `graph.print` statements under comments that people can uncomment to see how the graph is evolving after each statement

ansley replied:
Good point! I didn't want to clutter up the code for people who are only reading these examples, but I put a note in the docstring. I also added your `print_IR` method in a separate file, since I thought that might be helpful.

jamesr66a reviewed (approved):
Looks great, I added some more incremental improvement suggestions. Feel free to land when ready

Commit:
Update on "Document single op replacement"


Differential Revision: [D25803457](https://our.internmc.facebook.com/intern/diff/D25803457)

[ghstack-poisoned]
